cns/deviceplugin/server.go

```diff
@@ -0,0 +1,157 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const devicePrefix = "NIC-"
+
+type deviceCounter interface {
+	getDeviceCount() int
+}
+
+type Server struct {
+	address                  string
+	logger                   *zap.Logger
+	deviceCounter            deviceCounter
+	listAndWatchShutdownChan chan struct{}
+	deviceCheckInterval      time.Duration
+}
+
+func NewServer(logger *zap.Logger, address string, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:                  address,
+		logger:                   logger,
+		deviceCounter:            deviceCounter,
+		listAndWatchShutdownChan: make(chan struct{}),
+		deviceCheckInterval:      deviceCheckInterval,
+	}
+}
+
+// Run starts the grpc server and blocks until an error or context is cancelled. Wait on Ready to know when the server is ready.
+func (s *Server) Run(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	v1beta1.RegisterDevicePluginServer(grpcServer, s)
+	defer close(s.listAndWatchShutdownChan)
+
+	l, err := net.Listen("unix", s.address)
+	if err != nil {
+		return errors.Wrap(err, "error listening on socket")
+	}
+	defer l.Close()
+
+	go func() {
+		<-ctx.Done()
+		grpcServer.GracefulStop()
+	}()
+
+	if err := grpcServer.Serve(l); err != nil && errors.Is(err, grpc.ErrServerStopped) {
+		return errors.Wrap(err, "error running grpc server")
+	}
+	return nil
```

Comment:
do we only care about returning the error if it is `ErrServerStopped`? Seems like we would want to return any error from `Serve`?

Reply:
It was a typo, I meant to do below. corrected the code now
```
	if err := grpcServer.Serve(l); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
		return errors.Wrap(err, "error running grpc server")
	}
```
 I believe grpc.ErrServerStopped indicates the server has stopped normally, typically in response to a call to GracefulStop(). This is a part of the normal operation and shutdown sequence for gRPC servers, so it's not considered an error.